### PR TITLE
sqlstats: always enable tracing first time fingerprint is seen

### LINKED ISSUE #89185: Tracing should always log first time statement is executed
This [PR](https://github.com/cockroachdb/cockroach/pull/88343) turned off plan sampling. I side effect of this being disabled is the first time a statement is executed it is no longer being traced. This causes the contention event shows up in transaction_contention_events table to be empty most of the time. This was verified by running this to revert the flag `SET CLUSTER SETTING sql.metrics.statement_details.plan_collection.enabled = true;` and the contention event shows up in transaction_contention_events table. 

The PR to change the default was reverted, because of the impact. https://github.com/cockroachdb/cockroach/pull/89020

The logic needs to be fixed so it always traces the first time a statement is executed, so it does not depend on the above feature flag. The issue seems to be in [instrumentation.go](https://github.com/cockroachdb/cockroach/blob/ae38d903c527fdd85a0183b6245d191e51d79578/pkg/sql/instrumentation.go#L295).

Original PR always traced it: https://github.com/cockroachdb/cockroach/commit/125f6e6be9bf23bc4c69aff6abb02916a3fbd20d
Refactored here:
[https://github.com/cockroachdb/cockroach/commit/06f6874992c8358fef4baada3bec4c5bda359037#diff-a885af4d7f01dd0f13e3ab3[…]59f573ae9c8647aa60e218b0ba24](https://github.com/cockroachdb/cockroach/commit/06f6874992c8358fef4baada3bec4c5bda359037#diff-a885af4d7f01dd0f13e3ab3a77658f49acca59f573ae9c8647aa60e218b0ba24)

### PR DESCRIPTION
**Pull request by j82w**
Fixes: #89185

The first time a fingerprint is seen tracing should be enabled. This currently is broken if
sql.metrics.statement_details.plan_collection.enabled is set to false. This can cause crdb_internal.transaction_contention_events to be empty because tracing was never enabled to the contention event was never recorded.

To properly fix this a new value needs to be returned on ShouldSample to tell if it is the first time a fingerprint is seen. This will remove the dependency on plan_collection feature switch.

Release justification: Bug fixes and low-risk updates to new functionality.

Release note (bug fix): Always enable tracing the frist time a fingerprint is seen.

### CODE REVIEW

**Review by xinhaoz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89418)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @j82w and @yuzefovich)

___
*[`-- commits` line 10 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-NDdaToJE64K83D-1tfC):*
typo? to -> so
___
*[`-- commits` line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-NDdaYyb0CL8Sfta0VC7):*
typo frist -> first
___
*[`pkg/sql/instrumentation.go` line 258 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-NDdamO44nPQOOwKb-8T:-NDdamO44nPQOOwKb-8U:b-nnjpnd) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6c462ab841e6d2a682ce0f7583107cd6d7cd0a2/pkg/sql/instrumentation.go#L258)):*
> ```Go
> 	ih.withStatementTrace = cfg.TestingKnobs.WithStatementTrace
> 
> 	var everSampled bool
> ```

nit: just personal pref but this wording seems a little awkward to me. would something like `previouslySampled` or `sampleExists` be too lengthy? would just `sampled` suffice?
___
*[`pkg/sql/crdb_internal_test.go` line 941 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-NDdenW68YuKqQUF7Mfj:-NDdenW68YuKqQUF7Mfk:b-uqg98o) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6c462ab841e6d2a682ce0f7583107cd6d7cd0a2/pkg/sql/crdb_internal_test.go#L941)):*
> ```Go
> }
> 
> func TestTxnContentionEventsTable(t *testing.T) {
> ```

Is this possible as a data driven test?
___
*[`pkg/sql/crdb_internal_test.go` line 1038 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-NDde_j-63ci3jzdkFVH:-NDde_j-63ci3jzdkFVI:b-km8mvh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6c462ab841e6d2a682ce0f7583107cd6d7cd0a2/pkg/sql/crdb_internal_test.go#L1038)):*
> ```Go
> 
> 	require.Less(t, rowCount, 3, "node_execution_insights found 3 rows. It should only record first, "+
> 		"but there is a chance based ons sampling to get 2 rows.")
> ```

typo: ons -> on

**Review by yuzefovich — COMMENTED** [review]
Looks good overall! I left a bunch of nits, but please don't be discouraged by that - over time all of these code style things will become the second nature.

> Reviewed 9 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-:-NDfGsi56vd6K_byFwUc:buhf2m9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @j82w and @xinhaoz)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-NDfBXiU9BNTb35_s_7q):*
nit: use package name prefix to indicate where most of the changes are, also no capital letters in the commit message, so it should look something like `sqlstats: always enable ...`.
___
*[`-- commits` line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-NDfBqBX2IumfYu62rt4):*
nit: I don't think we've had a stable release with this bug, so I would omit the release note entirely. If you prefer to keep it, then I would explicitly mention that the bug would occur only if the plan collection setting is explicitly disabled by user. (Also, I believe we don't really document what and how we are sampling, which is yet another reason to not have a release note here.)
___
*[`pkg/sql/instrumentation.go` line 258 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-NDdamO44nPQOOwKb-8T:-NDfCGZO8BYKB_BUF1rA:bja8bya) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6c462ab841e6d2a682ce0f7583107cd6d7cd0a2/pkg/sql/instrumentation.go#L258)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

nit: just personal pref but this wording seems a little awkward to me. would something like `previouslySampled` or `sampleExists` be too lengthy? would just `sampled` suffice?
</blockquote></details>

I like `previouslySampled` more too.
___
*[`pkg/sql/sqlstats/sslocal/sslocal_stats_collector.go` line 129 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-NDfCVS95mzSbhJ5sgeY:-NDfCVS95mzSbhJ5sgeZ:b2gvp5a) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6c462ab841e6d2a682ce0f7583107cd6d7cd0a2/pkg/sql/sqlstats/sslocal/sslocal_stats_collector.go#L129)):*
> ```Go
> 	}
> 
> 	// TODO: this might need to be thought through a bit more.
> ```

We should address this TODO as part of this commit. I believe the current logic makes sense, but it'd be good for @xinhaoz to confirm it.
___
*[`pkg/sql/crdb_internal_test.go` line 941 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-NDdenW68YuKqQUF7Mfj:-NDfFEfRBHTX1ntw7BTA:bm6cw8s) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6c462ab841e6d2a682ce0f7583107cd6d7cd0a2/pkg/sql/crdb_internal_test.go#L941)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

Is this possible as a data driven test?
</blockquote></details>

It would probably be more involved to do it as a data-driven test, so I'd be in favor of keeping the current approach.

(If one is interested, we did add an ability for this kind of test to be done in the logic test framework - which is very similar to data driven framework - in #79010 (and maybe a couple of other related PRs from Alex Sarkesian) , and I would imagine we would have to do something similar here.)
___
*[`pkg/sql/crdb_internal_test.go` line 945 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-NDfFAug9wM-7E_ynyES:-NDfFAug9wM-7E_ynyET:bx3amyj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6c462ab841e6d2a682ce0f7583107cd6d7cd0a2/pkg/sql/crdb_internal_test.go#L945)):*
> ```Go
> 	defer log.Scope(t).Close(t)
> 
> 	// Start the cluster. (One node is sufficient; the outliers system is currently in-memory only.)
> ```

nit: we try to keep the comments wrapped at 80 characters. If you're using GoLand, you could set a visual guide (or automatic wrapping on typing) via Preferences -> Editor -> Code Style -> Go -> Wrapping and Braces.
___
*[`pkg/sql/crdb_internal_test.go` line 951 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-NDfFzqt03qn1iPYgSl-:-NDfFzqt03qn1iPYgSl0:b-lte76w) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6c462ab841e6d2a682ce0f7583107cd6d7cd0a2/pkg/sql/crdb_internal_test.go#L951)):*
> ```Go
> 	tc := testcluster.StartTestCluster(t, 1, args)
> 	defer tc.Stopper().Stop(ctx)
> 	conn := tc.ServerConn(0)
> ```

nit: you could wrap `conn` with `sqlutils.MakeSQLRunner` to avoid having to check errors and whatnot. This is just FYI, so please feel free to ignore this comment.
___
*[`pkg/sql/crdb_internal_test.go` line 975 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-NDfGo-WCcr1DHj_XV8v:-NDfGo-WCcr1DHj_XV8w:bxelaej) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6c462ab841e6d2a682ce0f7583107cd6d7cd0a2/pkg/sql/crdb_internal_test.go#L975)):*
> ```Go
> 		wgTxnStarted.Add(1)
> 
> 		// Lock to wait for the txn to complete to avoid the test finishing before the txn is committed
> ```

nit: missing period here and below.
___
*[`pkg/sql/crdb_internal_test.go` line 989 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-NDfFtLe5ua9DPN-jPrk:-NDfFtLe5ua9DPN-jPrl:biksn20) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6c462ab841e6d2a682ce0f7583107cd6d7cd0a2/pkg/sql/crdb_internal_test.go#L989)):*
> ```Go
> 			errTxn = tx.Commit()
> 			require.NoError(t, errTxn)
> 			wgTxnDone.Done()
> ```

nit: it's a good practice to do this in a `defer` that is the first line of the new goroutine, i.e. something like
```
waitGroup.Add(1)
go func() {
  defer waitGroup.Done()
  ...
}
```
___
*[`pkg/sql/sqlstats/persistedsqlstats/datadriven_test.go` line 132 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-NDfEZ-s-AbCluoy-lKc:-NDfEZ-s-AbCluoy-lKd:bxjqwhb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6c462ab841e6d2a682ce0f7583107cd6d7cd0a2/pkg/sql/sqlstats/persistedsqlstats/datadriven_test.go#L132)):*
> ```Go
> 			stubTime.setTime(tm)
> 			return stubTime.Now().String()
> 		case "should-sample-logical-plan":
> ```

nit: we probably should rename this to `should-sample`.
___
*[`pkg/sql/sqlstats/persistedsqlstats/datadriven_test.go` line 148 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-NDfERst1m56_Aqmt7KF:-NDfERst1m56_Aqmt7KG:b-fkyal5) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6c462ab841e6d2a682ce0f7583107cd6d7cd0a2/pkg/sql/sqlstats/persistedsqlstats/datadriven_test.go#L148)):*
> ```Go
> 			fingerprint = strings.Replace(fingerprint, "%", " ", -1)
> 
> 			// test harness needs an update.
> ```

nit: this comment should be removed.
___
*[`pkg/sql/sqlstats/persistedsqlstats/datadriven_test.go` line 154 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-NDfErdS9N37rJxs7_1a:-NDfErdS9N37rJxs7_1b:bq40d24) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6c462ab841e6d2a682ce0f7583107cd6d7cd0a2/pkg/sql/sqlstats/persistedsqlstats/datadriven_test.go#L154)):*
> ```Go
> 				dbName,
> 			)
> 			return fmt.Sprintf("%t, %t", savePlanForStats, everSampled)
> ```

nit: it seems more intuitive to keep the order of variables here in sync with `ShouldSample` - so either switch here or in the signature of `ShouldSample` itself.

**Comment by j82w** [comment]
*[`pkg/sql/crdb_internal_test.go` line 945 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-NDfFAug9wM-7E_ynyES:-NDhF8729C7DzfDhqjZh:bwrx82z) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6c462ab841e6d2a682ce0f7583107cd6d7cd0a2/pkg/sql/crdb_internal_test.go#L945)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: we try to keep the comments wrapped at 80 characters. If you're using GoLand, you could set a visual guide (or automatic wrapping on typing) via Preferences -> Editor -> Code Style -> Go -> Wrapping and Braces.
</blockquote></details>

I see in the coding standard is 80, but almost every file I go has some violation. The default is 120. I wonder if this could be enforced with a linter. https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/181371303/Go+Golang+coding+guidelines#Line-Length

**Review by xinhaoz — APPROVED** [review]
:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-:-NDhnIE_0lAvk1iRdGNt:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @j82w, @xinhaoz, and @yuzefovich)

___
*[`pkg/sql/sqlstats/sslocal/sslocal_stats_collector.go` line 123 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-NDhmZIAAvCaAoSgNPH5:-NDhmZIAAvCaAoSgNPH6:b-saj8b5) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dd7528feda60d19d2939d7941aeae39c7351519/pkg/sql/sqlstats/sslocal/sslocal_stats_collector.go#L123)):*
> ```Go
> ) (previouslySampled bool, savePlanForStats bool) {
> 	sampledInFlushTarget := false
> 	savePlanForStatsInFlushTarget := true
> ```

Thanks for this rename :thumbsup:

**Comment by j82w** [comment]
*[`pkg/sql/sqlstats/sslocal/sslocal_stats_collector.go` line 123 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-NDhmZIAAvCaAoSgNPH5:-NDi8gjr4rVo_-M1p6yK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dd7528feda60d19d2939d7941aeae39c7351519/pkg/sql/sqlstats/sslocal/sslocal_stats_collector.go#L123)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

Thanks for this rename :thumbsup: 
</blockquote></details>

Done.

**Review by yuzefovich — COMMENTED** [review]
:lgtm:

> Reviewed 4 of 7 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-:-NDiSpxd9DLKRow0f83Y:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @j82w and @xinhaoz)

___
*[`pkg/sql/sqlstats/sslocal/sslocal_stats_collector.go` line 129 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-NDfCVS95mzSbhJ5sgeY:-NDiRu2LDU1Qz8qBuB7g:bnk51b) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6c462ab841e6d2a682ce0f7583107cd6d7cd0a2/pkg/sql/sqlstats/sslocal/sslocal_stats_collector.go#L129)):*
<details><summary><i>Previously, j82w wrote…</i></summary><blockquote>

The only concern I have is when the stats get flushed it will see all statements as new again.
</blockquote></details>

Hm, I see. How frequent are these flushes? I'll defer to you and Xin Hao on whether this is concerning. I guess you could leave a TODO to address it later (just use a github username for it, i.e. something like `TODO(j82w): ...`).
___
*[`pkg/sql/crdb_internal_test.go` line 945 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-NDfFAug9wM-7E_ynyES:-NDiSH_3BmrtsnXXmHZU:bsw1jhg) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6c462ab841e6d2a682ce0f7583107cd6d7cd0a2/pkg/sql/crdb_internal_test.go#L945)):*
<details><summary><i>Previously, j82w wrote…</i></summary><blockquote>

I see in the coding standard is 80, but almost every file I go has some violation. The default is 120. I wonder if this could be enforced with a linter. https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/181371303/Go+Golang+coding+guidelines#Line-Length 
</blockquote></details>

Yes, we definitely have many violations, still we should try to keep them at minimum. We are also more stringent about the character count for comments than for the regular code.
___
*[`pkg/sql/crdb_internal_test.go` line 1019 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-NDiSa3SAxIYAvEoTysC:-NDiSa3SAxIYAvEoTysD:bn1whko) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dd7528feda60d19d2939d7941aeae39c7351519/pkg/sql/crdb_internal_test.go#L1019)):*
> ```Go
> 	// internal workflow.
> 	testutils.SucceedsWithin(t, func() error {
> 		rows, errVerify := conn.QueryContext(ctx, "SELECT "+
> ```

nit: with multiple-line single SQL statements we usually use backticks `. Feel free to ignore this nit, just FYI.

**Comment by xinhaoz** [comment]
*[`pkg/sql/sqlstats/sslocal/sslocal_stats_collector.go` line 129 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-NDfCVS95mzSbhJ5sgeY:-NDiYn5W1bbqzZQop9-U:b-astpgh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6c462ab841e6d2a682ce0f7583107cd6d7cd0a2/pkg/sql/sqlstats/sslocal/sslocal_stats_collector.go#L129)):*
> The only concern I have is when the stats get flushed it will see all statements as new again.

The flushes are default every 10mins, but we were already seeing the behaviour described by Jake prior to this change.

**Comment by j82w** [comment]
*[`pkg/sql/sqlstats/sslocal/sslocal_stats_collector.go` line 129 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89418#-NDfCVS95mzSbhJ5sgeY:-NE1ZI5q3VH5fzNcY9x3:b-r226ss) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6c462ab841e6d2a682ce0f7583107cd6d7cd0a2/pkg/sql/sqlstats/sslocal/sslocal_stats_collector.go#L129)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

> The only concern I have is when the stats get flushed it will see all statements as new again.

The flushes are default every 10mins, but we were already seeing the behaviour described by Jake prior to this change.
</blockquote></details>

I confirmed that it currently always records the statement if you wait 10mins. In that case I will not add a todo note, because it's the current behavior. We should wait for either customer feedback or perf analysis to show it's an issue.

**Pull request merged**
